docs/proposals/container-runtime-interface-v1.md

```diff
@@ -0,0 +1,255 @@
+in the sandbox before deleting the sandbox. If Kubelet chooses to delete
+the sandbox with running containers in it, those containers should be forcibly
+deleted.
+
+Note that every PodSandbox/container lifecycle operation (create, start,
+stop, delete) should either return an error or block until the operation
+succeeds. A successful operation should include a state transition of the
+PodSandbox/container. E.g., if a `Create` call for a container does not
+return an error, the container state should be "created" when the runtime is
+queried.
+
+### Updates to PodSandbox or Containers
+
+Kubernetes support updates only to a very limited set of fields in the Pod
+Spec.  These updates may require containers to be re-created by Kubelet. This
+can be achieved through the proposed, imperative container-level interface.
+On the other hand, PodSandbox update currently is not required.
+
+
+### Container Lifecycle Hooks
+
+Kubernetes supports post-start and pre-stop lifecycle hooks, with ongoing
+discussion for supporting pre-start and post-stop hooks in
+[#140](https://issues.k8s.io/140).
+
+These lifecycle hooks will be implemented by Kubelet via `Exec` calls to the
+container runtime. This frees the runtimes from having to support hooks
+natively.
+
+Illustration of the container lifecycle and hooks:
+
+```
+            pre-start post-start    pre-stop post-stop
+               |        |              |       |
+              exec     exec           exec    exec
+               |        |              |       |
+ create --------> start ----------------> stop --------> remove
+```
+
+In order for the lifecycle hooks to function as expected, the `Exec` call
+will need access to the container's filesystem (e.g., mount namespaces).
+
+### Extensibility
+
+There are several dimensions for container runtime extensibility.
+ - Host OS (e.g., Linux)
+ - PodSandbox isolation mechanism (e.g., namespaces or VM)
+ - PodSandbox OS (e.g., Linux)
+
+As mentioned previously, this proposal will only address the Linux based
+PodSandbox and containers. All Linux-specific configuration will be grouped
+into one field. A container runtime is required to enforce all configuration
+applicable to its platform, and should return an error otherwise.
+
+## Alternatives
+
+**[Status quo] Declarative pod-level interface**
+ - Pros: No changes needed.
+ - Cons: All the issues stated in #motivation
+
+**Allow integration at both pod- and container-level interfaces**
+ - Pros: Flexibility.
+ - Cons: All the issues stated in #motivation
+
+**Imperative pod-level interface**
+The interface contains only CreatePod(), StartPod(), StopPod() and RemovePod().
+ - Pros: Kubelet can potentially become a very thin daemon; lower maintenance
+    overhead for the Kubernetes maintainers.
+ - Cons: Higher integration cost and lower feature velocity.
+
+## Related Issues
+
+ * `AttachContainer`: [#23335](https://issues.k8s.io/23335)
```

Comment:
Sgtm.  Key concern is that this feature should not be read as a plan to
deprecate those two pieces of function.

On Wednesday, June 8, 2016, Yu-Ju Hong <notifications@github.com> wrote:

> In docs/proposals/container-runtime-interface-v1.md
> <https://github.com/kubernetes/kubernetes/pull/25899#discussion_r66373889>
> :
>
> > + - Pros: No changes needed.
> > + - Cons: All the issues stated in #motivation
> > +
> > +**Allow integration at both pod- and container-level interfaces**
> > + - Pros: Flexibility.
> > + - Cons: All the issues stated in #motivation
> > +
> > +**Imperative pod-level interface**
> > +The interface contains only CreatePod(), StartPod(), StopPod() and RemovePod().
> > + - Pros: Kubelet can potentially become a very thin daemon; lower maintenance
> > +    overhead for the Kubernetes maintainers.
> > + - Cons: Higher integration cost and lower feature velocity.
> > +
> > +## Related Issues
> > +
> > + * `AttachContainer`: [#23335](https://issues.k8s.io/23335)
>
> @derekwaynecarr <https://github.com/derekwaynecarr>, @vishh
> <https://github.com/vishh> and I discussed today and we decided to leave
> out "uncertain" features in the first draft since this interface is in its
> early stage, and we want to keep it minimal. I've updated the proposal to
> reflect this. We can add them later if needed. Also, this shouldn't affect
> existing users (with docker or rkt) since we can always plumb them
> through internally in our codebase.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25899/files/34028f5148072d348855d80bf701710c97ac308d#r66373889>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AF8dbHw1d4tiA0ZCBdV-tHw8zov0PICZks5qJ3iOgaJpZM4Iil-->
> .
>